File: 07.py
#!/usr/bin/env python

from parse import parse


def pairs():
    with open("07.input") as f:
        for line in f:
            a, b = parse(
                'Step {} must be finished before step {} can begin.', line)
            yield (a, b)


class Jobs:
    def __init__(self, pairs):
        self.d = {}
        for a, b in pairs:
            self.d.setdefault(a, set())
            self.d.setdefault(b, set()).add(a)

    def __len__(self):
        return len(self.d)

    def ready(self):
        '''Jobs that have no pending dependencies.'''
        for k, v in self.d.items():
            if not v:
                yield k

    def start(self, job):
        '''Start working on a job. It will be removed from the ready set.'''
        assert job in self.d
        del self.d[job]
        return job

    def finish(self, job):
        '''Finish working on a job. It will be removed from others' deps.'''
        assert job not in self.d
        for deps in self.d.values():
            deps.discard(job)


class Crew:
    def __init__(self, size, jobs):
        self.t = 0
        self.workers = [None] * size
        self.jobs = jobs
        self.done = []

    def __str__(self):
        workset = ' '.join(w[0] if w else '.' for w in self.workers)
        return "{:3}: {} | {}".format(self.t, workset, ''.join(self.done))

    def job_time(self, job):
        return 60 + 1 + ord(job) - ord('A')

    def handoff(self):
        for i, w in enumerate(self.workers):
            if w is None:
                continue
            job, remain = w
            if remain == 0:
                self.workers[i] = None
                self.jobs.finish(job)
                self.done.append(job)

    def pickup(self):
        ready = sorted(self.jobs.ready())
        for i, w in enumerate(self.workers):
            if not ready:
                break
            if w is None:
                job = ready.pop(0)
                remain = self.job_time(job)
                self.workers[i] = (job, remain)
                self.jobs.start(job)

    def tick(self):
        self.t += 1
        for i, w in enumerate(self.workers):
            if w is None:
                continue
            job, remain = w
            self.workers[i] = (job, remain - 1)

    def step(self):
        self.pickup()
        self.tick()
        print(self)
        self.handoff()

    def run(self):
        while self.jobs or any(self.workers):
            self.step()
        return self.t, self.done


if __name__ == '__main__':
    # part 1
    jobs = Jobs(pairs())
    while jobs:
        job = sorted(jobs.ready())[0]
        jobs.start(job)
        jobs.finish(job)
        print(job, end='')
    print()

    # part 2
    crew = Crew(5, Jobs(pairs()))
    t, done = crew.run()
    print(t, ''.join(done))

File: 07.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.10 script --package containers

{-# OPTIONS_GHC -Wall #-}

import Data.Char (ord)
import Data.Map (Map)
import qualified Data.Map as Map
import Data.Maybe (fromMaybe, listToMaybe, maybeToList)
import Text.ParserCombinators.ReadP
import Utils

type Step = Char
type Dep = (Step, Step)  -- fst must finish before snd can begin
type Deps = Map Step [Step] -- key cannot start until [val] are finished

dep :: ReadP Dep
dep = do
    a <- string "Step " >> get
    b <- string " must be finished before step " >> get <*
         string " can begin." <* eof
    return (a, b)

-- Convert (a, b) pairs into mapping b -> [a]
buildDeps :: [Dep] -> Deps
buildDeps = go Map.empty where
    go m [] = m
    go m ((a, b):os) = go addDep os where
        addDep = Map.insertWith (++) b [a] $ Map.insertWith (++) a [] m

-- Return steps that have zero dependencies
avail :: Deps -> [Step]
avail = Map.keys . fst . Map.partition null

-- Return next available step (using alphabet to break ties)
nextAvail :: Deps -> Maybe Step
nextAvail = listToMaybe . avail

-- Start a job by removing it from the map keys, i.e. no longer available
startJob :: Step -> Deps -> Deps
startJob = Map.delete

-- Finish a job by removing it from other jobs' dependencies
finishJob :: Step -> Deps -> Deps
finishJob s = Map.map (filter (/=s))

-- Return all steps in sorted order
inOrder :: Deps -> [Step]
inOrder = go [] where
    go finished jobs = case nextAvail jobs of
        Nothing -> reverse finished
        Just s -> go (s:finished) jobs' where
            jobs' = finishJob s $ startJob s jobs

-- How long does it take to complete a job
workTime :: Step -> Int
workTime c = 60 + ord c - (ord 'A' - 1)

data Worker = Worker {
    job :: Maybe Step,
    remains :: Int }
    deriving (Eq, Show)
data State = State {
    t :: Int,
    todo :: Deps,
    done :: [Step],
    workers :: [Worker] }
    deriving Show

idleWorker :: Worker
idleWorker = Worker Nothing 0

-- Return next state of a Worker, deliver finished jobs and return to idle
work2idle :: (Worker, Deps) -> (Worker, Deps, Maybe Step)
work2idle (Worker Nothing _, jobs) = (idleWorker, jobs, Nothing) -- still idle
work2idle (Worker (Just job') remains', jobs) =
    let remains'' = remains' - 1 in
    if remains'' > 0
    then (Worker (Just job') remains'', jobs, Nothing) -- still working
    else (idleWorker, finishJob job' jobs, Just job') -- finished

-- Return next state of a Worker, picking up new jobs if idle
idle2work :: (Worker, Deps) -> (Worker, Deps)
idle2work (Worker Nothing _, jobs) =
    maybe (idleWorker, jobs) begin (nextAvail jobs) where
        begin newjob = (newWorker, startJob newjob jobs) where
            newWorker = Worker (Just newjob) (workTime newjob)
idle2work (w, jobs) = (w, jobs) -- already busy

-- Deliver finished jobs
handoff :: (Deps, [Worker]) -> (Deps, [Worker], [Step])
handoff (jobs, ws) = go jobs ws [] [] where
    go jobs' [] reported finished = (jobs', reported, finished)
    go jobs' (w:ws') reported finished =
        go jobs'' ws' (w':reported) finished' where
            (w', jobs'', maybeDone) = work2idle (w, jobs')
            finished' = maybeToList maybeDone ++ finished

-- Pick up new jobs
pickup :: (Deps, [Worker]) -> (Deps, [Worker])
pickup (jobs, ws) = go jobs ws [] where
    go jobs' [] assigned = (jobs', assigned)
    go jobs' (w:ws') assigned =
        go jobs'' ws' (w':assigned) where
            (w', jobs'') = idle2work (w, jobs')

-- Daily standup for the workers, advance each of them, and thread Deps through
standup :: (Deps, [Worker]) -> (Deps, [Worker], [Step])
standup (jobs, ws) = (jobs'', ws'', finished) where
    (jobs', ws', finished) = handoff (jobs, ws)
    (jobs'', ws'') = pickup (jobs', ws')

-- Return state at t0, workers have just picked up their first jobs
initState :: Int -> Deps -> State
initState numWorkers jobs =
    State { t = 0, todo = jobs', done = [], workers = initWorkers } where
        idleWorkers = replicate numWorkers idleWorker
        (jobs', initWorkers) = pickup (jobs, idleWorkers)

-- Advance one time unit and return updated state
next :: State -> State
next (State t' jobs finished ws) =
    State { t = t' + 1, todo = jobs', done = done', workers = ws' } where
        (jobs', ws', nowDone) = standup (jobs, ws)
        done' = nowDone ++ finished

-- Tick over the clock until all workers are idle
deathMarch :: State -> IO State
deathMarch s =
    if all (==idleWorker) (workers s)
    then return s
    else do
        putStr $ showState s
        deathMarch $ next s

-- Show current state
showState :: State -> String
showState s = show (t s) ++ ":" ++ workerStr ++ " | " ++ doneStr ++ "\n" where
        workerStr = concatMap (\w -> [' ', fromMaybe '.' (job w)]) (workers s)
        doneStr = reverse (done s)

-- Pipe output through 'dot -Txlib'
dot :: [Dep] -> String
dot o = "digraph ordering {\n" ++ concatMap go o ++ "}\n" where
    go (a, b) = "    " ++ [a] ++ " -> " ++ [b] ++ "\n"

main :: IO ()
main = do
    input <- readFile "07.input"
    let deps = parseMany dep $ lines input
    let jobs = buildDeps deps
    -- part 1
    print $ inOrder jobs
    -- part 2
    finished <- deathMarch $ initState 5 jobs
    print $ t finished
